contracts/staking-vault/StakingVault.sol

```diff
@@ -0,0 +1,284 @@
+
+    return ((sTokenPrice[tokenID] * amountBurnt) / 1e18) / 10**(18 - decimals);
+  }
+
+  function addRewards(address tokenBacked, uint256 rewardAmount) external {
+    _addRewards(rewardAmount, tokenBacked);
+    IERC20(rewardToken).safeTransferFrom(msg.sender, address(this), rewardAmount);
+
+    emit AddRewards(tokenBacked, rewardAmount);
+  }
+
+  function accumulateRewards(address tokenBacked, uint256 rewardAmount) external {
+    undistributedRewards[tokenBacked] += rewardAmount;
+    IERC20(rewardToken).safeTransferFrom(msg.sender, address(this), rewardAmount);
+
+    emit AccumulateRewards(tokenBacked, rewardAmount);
+  }
+
+  function distributeRewards(address tokenBacked) external {
+    uint256 rewards = undistributedRewards[tokenBacked];
+    _addRewards(rewards, tokenBacked);
+    undistributedRewards[tokenBacked] = 0;
+
+    emit DistributeRewards(tokenBacked, rewards);
+  }
+
+  function _addRewards(uint256 rewardAmount, address tokenBacked) internal {
+    address[] memory tokensStaked = stakingTokens;
+    uint256 totalValue;
+    uint256[] memory values = new uint256[](tokensStaked.length);
+    uint256 tokenID;
+    for (uint256 i; i < tokensStaked.length; ) {
+      tokenID = convertToID(tokensStaked[i], tokenBacked);
+      values[i] = priceFeed.queryReturn(tokensStaked[i], valuationToken, balanceStakedPerID[tokenID]);
+      totalValue += values[i];
+      unchecked {
+        ++i;
+      }
+    }
+    for (uint256 i; i < tokensStaked.length; ) {
+      tokenID = convertToID(tokensStaked[i], tokenBacked);
+      values[i] = (values[i] * 1e18) / totalValue;
+      rewardsPerToken[tokenID] += (rewardAmount * values[i]) / supplyPerID[tokenID];
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function claimRewards(uint256[] memory tokenIDs) external {
+    for (uint256 i; i < tokenIDs.length; ) {
+      _claimReward(tokenIDs[i]);
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function _claimReward(uint256 tokenID) internal {
+    uint256 previousAmount = lastClaimRewardAccrual[msg.sender][tokenID];
+    if (previousAmount == 0 && initialized[tokenID]) {
+      lastClaimRewardAccrual[msg.sender][tokenID] = rewardsPerToken[tokenID];
+      return;
+    }
+    uint256 newAmount = rewardsPerToken[tokenID];
+    if (newAmount - previousAmount == 0) {
+      return;
+    }
+    IERC20(rewardToken).safeTransfer(msg.sender, ((newAmount - previousAmount) * balanceOf(msg.sender, tokenID)) / 1e18);
+    if (!initialized[tokenID] && newAmount > 0) {
+      initialized[tokenID] = true;
+    }
+    lastClaimRewardAccrual[msg.sender][tokenID] = newAmount;
+  }
+
+  function setPriceFeed(IPriceFeeds p) external onlyOwner {
+    priceFeed = p;
+
+    emit PriceFeed(address(p));
+  }
+
+  function setProtocol(address p) external onlyOwner {
+    protocol = p;
+
+    emit Protocol(p);
+  }
+
+  function setValuationToken(address v) external onlyOwner {
+    valuationToken = v;
```

Comment:
should we have this as a constant instead? immutable?

Reply:
also protocol and pricefeed

Reply:
I think this is a good idea.

Reply:
implemented